core/src/io/anuke/mindustry/world/blocks/power/PowerGraph.java

```diff
@@ -31,12 +35,62 @@ public int getID(){
     }
 
     public void update(){
-        if(threads.getFrameID() == lastFrameUpdated || consumers.size == 0 || producers.size == 0){
+        if(threads.getFrameID() == lastFrameUpdated || (consumers.size == 0 && producers.size == 0) ||
+                (consumers.size == 0 && batteries.size == 0) || (producers.size == 0 && batteries.size == 0))  {
             return;
         }
 
         lastFrameUpdated = threads.getFrameID();
 
+        float totalProduction = 0f;
+        for(Tile producer : producers){
+            totalProduction += producer.block().powerProduction * producer.entity.power.productionEfficiency * Timers.delta();
+        }
+
+        float requiredPower = 0f;
+        for(Tile consumer : consumers){
+            requiredPower += consumer.block().powerCapacity - consumer.entity.power.buffer;
+        }
+
+        float totalAccumulator = 0f;
+        float totalStorage = 0f;
+        for (Tile battery: batteries) {
+            totalAccumulator += battery.entity.power.buffer;
+            totalStorage += battery.block().powerCapacity - battery.entity.power.buffer;
+        }
+
+        if ((totalProduction <= 0.0001f && requiredPower <= 0.0001f) ||
+                (totalProduction <= 0.0001f && totalAccumulator <= 0.0001f) || (requiredPower <= 0.0001f && totalStorage <= 0.0001f)) {
+            return;
+        }
+
+        if (totalProduction < requiredPower && totalAccumulator > 0.0001f) {
+            float accumulatorUsed = Math.min(requiredPower - totalProduction, totalAccumulator);
+            totalProduction += accumulatorUsed;
+            for (Tile battery: batteries) {
+                battery.entity.power.buffer *= 1f - (accumulatorUsed / totalAccumulator);
+            }
+        }
+
+        Array<Tile> sortedConsumers = consumers.iterator().toArray();
```

Comment:
If you're calling this *every frame*, it's going to pollute the heap with new arrays.
Use an `OrderedSet<Tile>` for consumers, iterate through it using `consumers.orderedItems()` and sort it with `consumers.orderedItems().sort(...)` *only when a new consumer is added*.

Reply:
The order has to be updated every frame because consumers still use buffers so they can change every frame, so I think I'll revert to my previous algorithm.